pkg/streamer/reader.go

```diff
@@ -103,9 +117,203 @@ func (r *BinlogReader) checkRelayPos(pos mysql.Position) error {
+	ctx2, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	s := newLocalStreamer()
+	parser := replication.NewBinlogParser()
+	parser.SetVerifyChecksum(true)
+	parser.SetUseDecimal(true)
+	if r.cfg.Timezone != nil {
+		parser.SetTimestampStringLocation(r.cfg.Timezone)
+	}
+
+	r.wg.Add(1)
+	go func() {
+		defer r.wg.Done()
+
+		offset := int64(4)
+		onEventFunc := func(e *replication.BinlogEvent) error {
+			offset = int64(e.Header.LogPos)
+
+			select {
+			case s.ch <- e:
+			case <-ctx2.Done():
+			}
+			return nil
+		}
+
+		err := parser.ParseFile(filePath, offset, onEventFunc)
+		if err != nil {
+			r.tctx.L().Error("parse file stopped", zap.Error(err))
+			select {
+			case s.ech <- err:
+			case <-ctx2.Done():
+			}
+		} else {
+			// reach end of file
+			select {
+			case s.ch <- nil:
+			case <-ctx2.Done():
+			}
+		}
+	}()
+
+	lastPos := uint32(0)
+	for {
+		select {
+		case <-ctx.Done():
+			return 0, nil
+		default:
+		}
+
+		e, err := s.GetEvent(ctx)
+		if err != nil {
+			return 0, err
+		}
+		// gset is the last txn in file
+		if e == nil {
+			return lastPos, nil
+		}
+		switch ev := e.Event.(type) {
+		case *replication.PreviousGTIDsEvent:
+			// nil previous gtid event
+			if len(ev.GTIDSets) == 0 {
+				lastPos = e.Header.LogPos
+				continue
+			}
+			gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, ev.GTIDSets)
+			if err != nil {
+				return 0, err
+			}
+			// if PreviousGITDsEvent contain the gset, go to previous file
+			if gs.Contain(gset) {
+				return 0, nil
+			}
+		case *replication.RotateEvent:
+			if e.Header.Timestamp != 0 && e.Header.LogPos != 0 {
+				// gset is the last txn in file
+				return lastPos, nil
+			}
+		case *replication.XIDEvent:
+			lastPos = e.Header.LogPos
+		case *replication.QueryEvent:
+			lastPos = e.Header.LogPos
+		case *replication.GTIDEvent:
+			u, _ := uuid.FromBytes(ev.SID)
+			gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, fmt.Sprintf("%s:%d", u.String(), ev.GNO))
+			if err != nil {
+				return 0, err
+			}
+			// meet first gtid event greater than gset
+			if !gset.Contain(gs) {
+				return lastPos, nil
+			}
+		case *replication.MariadbGTIDEvent:
+			GTID := ev.GTID
+			gs, err := mysql.ParseGTIDSet(r.cfg.Flavor, fmt.Sprintf("%d-%d-%d", GTID.DomainID, GTID.ServerID, GTID.SequenceNumber))
+			if err != nil {
+				return 0, err
+			}
+			// meet first gtid event greater than gset
+			if !gset.Contain(gs) {
+				return lastPos, nil
+			}
+		default:
+		}
+	}
+}
+
+// getPosByGTID gets position by gtid
+func (r *BinlogReader) getPosByGTID(gset mysql.GTIDSet) (*mysql.Position, error) {
+	uuid, err := r.getUUIDByGTID(gset)
+	if err != nil {
+		return nil, err
+	}
+	_, suffix, err := utils.ParseSuffixForUUID(uuid)
+	if err != nil {
+		return nil, err
+	}
+
+	uuidDir := path.Join(r.cfg.RelayDir, uuid)
+	allFiles, err := CollectAllBinlogFiles(uuidDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// iterate files from the newest one
```

Comment:
in future we could use a binary search to find "the last file returns non-zero `pos`" and use that `pos`